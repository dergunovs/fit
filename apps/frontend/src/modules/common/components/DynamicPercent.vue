<template>
  <sup
    :class="$style.dynamics"
    :data-positive="(props.percent || 0) > 0"
    :data-negative="(props.percent || 0) < 0"
    data-test="dynamic-percent"
  >
    {{ formatPercent(props.percent || 0) }}
  </sup>
</template>

<script setup lang="ts">
import { formatPercent } from 'mhz-helpers';

interface IProps {
  percent: number | null;
}

const props = defineProps<IProps>();
</script>

<style module lang="scss">
.dynamics {
  font-size: 0.75rem;
  color: var(--color-gray-dark);

  &[data-positive='true'] {
    color: var(--color-success-dark);
  }

  &[data-negative='true'] {
    font-weight: 700;
    color: var(--color-error);
  }
}
</style>
