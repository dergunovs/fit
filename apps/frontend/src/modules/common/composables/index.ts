import { ref, computed, onMounted } from 'vue';
import { useRouter, useRoute } from 'vue-router';

import LayoutAdmin from '@/common/components/LayoutAdmin.vue';
import LayoutDefault from '@/common/components/LayoutDefault.vue';

export function useLayout() {
  const router = useRouter();
  const route = useRoute();

  const isLoaded = ref(false);

  const layoutComponent = computed(() => (route.meta.layout === 'admin' ? LayoutAdmin : LayoutDefault));

  onMounted(async () => {
    await router.isReady();
    isLoaded.value = true;
  });

  return {
    isLoaded,
    layoutComponent,
  };
}

export function useRouteId(title: string, isQuery?: boolean) {
  const route = useRoute();

  const id = computed(() => {
    const value = isQuery ? route.query[title]?.toString() : route.params[title].toString();

    return value || '';
  });

  return { id };
}
