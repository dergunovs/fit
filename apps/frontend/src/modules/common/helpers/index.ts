import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
