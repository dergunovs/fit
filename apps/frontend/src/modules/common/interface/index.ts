import { TPostAuthLoginDTO } from 'fitness-tracker-contracts';
import { FunctionalComponent } from 'vue';

export interface INavItem {
  url: string;
  title: string;
  icon: FunctionalComponent;
}

export interface IOnSuccess {
  create?: (id?: string) => Promise<void>;
  update?: () => Promise<void>;
  delete?: () => Promise<void>;
  login?: (user: TPostAuthLoginDTO) => Promise<void>;
  setup?: () => Promise<void>;
  register?: () => Promise<void>;
}

export interface IPromptEvent extends Event {
  prompt: () => Promise<void>;
}

export interface IGetLatestVersionDTO {
  'dist-tags': { latest: string };
}
