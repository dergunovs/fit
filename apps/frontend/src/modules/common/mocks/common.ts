import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
