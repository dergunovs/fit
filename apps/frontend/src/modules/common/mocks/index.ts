export * from './helpers';
export * from './ui';
export * from './router';
export * from './layout';
export * from './pwa';
export * from './service';
