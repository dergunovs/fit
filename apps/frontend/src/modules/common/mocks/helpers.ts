import { computed, ref, Ref } from 'vue';
import { vi } from 'vitest';
import type { UseQueryReturnType, UseMutationReturnType } from 'mhz-helpers';
import * as helpers from 'mhz-helpers';

import * as commonHelpers from '@/common/helpers';

const mockDateFrom = ref('2024-11-24T21:00:00.000Z');
const mockDateTo = ref('2025-01-05T20:59:59.000Z');
const mockIsDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyCopyToClipboard = vi.spyOn(commonHelpers, 'copyToClipboard');

function returnComputed<T>(value: T) {
  return computed(() => value);
}

const mockIsValid = ref(true);

vi.spyOn(helpers, 'useValidator').mockImplementation(() => {
  return { error: () => undefined, isValid: () => mockIsValid.value, errors: returnComputed(undefined) };
});

const spyUseCalendar = vi.spyOn(helpers, 'useCalendar').mockReturnValue({
  dateFrom: mockDateFrom,
  dateTo: mockDateTo,
  isDatesReady: mockIsDatesReady,
  updateDates: spyUpdateDates,
});

const spyUsePageLock = vi.spyOn(helpers, 'usePageLock');

const mockTempId = 'fd98bye9rbrube';

vi.spyOn(helpers, 'createTempId').mockReturnValue(mockTempId);

const spyAuth = vi.fn();

vi.spyOn(helpers, 'useAuth').mockReturnValue({ auth: spyAuth, redirectIfAuth: () => undefined });

const spySetAuthHeaders = vi.spyOn(helpers, 'setAuthHeader');
const spyLogout = vi.spyOn(helpers, 'logout');

const mockPageNumber = ref(1);
const spySetPage = vi.fn();

const spyUsePageNumber = vi.spyOn(helpers, 'usePageNumber').mockReturnValue({
  page: mockPageNumber,
  setPage: spySetPage,
  resetPage: () => undefined,
});

const spyUsePagination = vi
  .spyOn(helpers, 'usePagination')
  .mockImplementation(<T>(dataRaw: Ref<{ data: T[]; total: number } | undefined>) => {
    return {
      data: returnComputed(dataRaw.value?.data),
      total: returnComputed(dataRaw.value?.total),
      setPaginationPage: () => 1,
    };
  });

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const mockRouteId = computed(() => '123');

const spyUseRouteId = vi.spyOn(helpers, 'useRouteId').mockImplementation(() => {
  return { id: mockRouteId };
});

function mockQueryReply<T>(reply: object | string, refetch?: () => void) {
  return { data: ref(reply), refetch, isSuccess: true } as unknown as UseQueryReturnType<T, Error>;
}

function mockMutationReply<T, T2>(mutate: () => void) {
  return { mutate } as unknown as UseMutationReturnType<T, Error, T2, unknown>;
}

export {
  spyCopyToClipboard,
  spyAuth,
  spySetAuthHeaders,
  spyLogout,
  spyUsePageNumber,
  spyUsePagination,
  spySetPage,
  spyRefetchQueries,
  spyRemoveQueries,
  spyUpdateDates,
  spyUseCalendar,
  spyUsePageLock,
  spyUseRouteId,
  mockDateFrom,
  mockDateTo,
  mockIsDatesReady,
  mockIsValid,
  mockQueryReply,
  mockMutationReply,
  mockTempId,
  mockPageNumber,
  mockRouteId,
};
