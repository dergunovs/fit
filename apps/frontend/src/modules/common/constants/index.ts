import IconExercise from '@/common/icons/nav-exercise.svg?component';
import IconActivity from '@/common/icons/nav-activity.svg?component';
import IconEquipment from '@/common/icons/weight.svg?component';
import IconUser from '@/common/icons/nav-user.svg?component';

import { URL_USER } from '@/user/constants';
import { URL_EXERCISE } from '@/exercise/constants';
import { URL_ACTIVITY_ADMIN } from '@/activity/constants';
import { URL_EQUIPMENT } from '@/equipment/constants';
import { INavItem } from '@/common/interface';
import { URL_SETUP } from '@/auth/constants';

export const NAV_ITEMS: INavItem[] = [
  { _id: '1', url: URL_EXERCISE, title: 'Упражнения', icon: IconExercise },
  { _id: '2', url: URL_ACTIVITY_ADMIN, title: 'Занятия', icon: IconActivity },
  { _id: '3', url: URL_EQUIPMENT, title: 'Оборудование', icon: IconEquipment },
  { _id: '4', url: URL_USER, title: 'Пользователи', icon: IconUser },
];

export const UPDATE_BUTTON_TEXT = 'Обновить';
export const CREATE_BUTTON_TEXT = 'Создать';

export const URL_HOME = '/';
export const URL_ERROR = '/404';

export const GUEST_PAGES = [URL_HOME, URL_SETUP];
