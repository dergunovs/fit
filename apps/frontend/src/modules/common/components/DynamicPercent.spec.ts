import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import DynamicPercent from './DynamicPercent.vue';

import { dataTest, wrapperFactory } from '@/common/test';
import { formatStatisticsPercent } from '@/common/helpers';

const dynamicPercent = dataTest('dynamic-percent');
const percent = 0;

let wrapper: VueWrapper<InstanceType<typeof DynamicPercent>>;

beforeEach(() => {
  wrapper = wrapperFactory(DynamicPercent, { percent });
});

enableAutoUnmount(afterEach);

describe('DynamicPercent', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(DynamicPercent)).toBeTruthy();
  });

  it('shows formatted percent', async () => {
    expect(wrapper.find(dynamicPercent).text()).toBe(formatStatisticsPercent(percent));
  });

  it('adds styling if percent is positive or negative', async () => {
    expect(wrapper.find(dynamicPercent).attributes('data-positive')).toBe('false');
    expect(wrapper.find(dynamicPercent).attributes('data-negative')).toBe('false');

    await wrapper.setProps({ percent: -2 });

    expect(wrapper.find(dynamicPercent).attributes('data-positive')).toBe('false');
    expect(wrapper.find(dynamicPercent).attributes('data-negative')).toBe('true');

    await wrapper.setProps({ percent: 2 });

    expect(wrapper.find(dynamicPercent).attributes('data-positive')).toBe('true');
    expect(wrapper.find(dynamicPercent).attributes('data-negative')).toBe('false');
  });
});
