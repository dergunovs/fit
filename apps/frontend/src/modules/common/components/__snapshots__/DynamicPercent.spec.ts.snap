// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DynamicPercent > matches snapshot 1`] = `"<sup class="_dynamics_186d7f" data-positive="false" data-negative="false" data-test="dynamic-percent">0%</sup>"`;
